config.go

```diff
@@ -34,6 +35,9 @@ type ClientConfig struct {
 	AzureModelMapperFunc func(model string) string // replace model to azure deployment name func
 	HTTPClient           *http.Client
 
+	// allows additional modification of request before sending
+	CustomRequestModifier func(ctx context.Context, req *http.Request) error
```

Comment:
@arrix  Thank you for the suggestion. However, this can already be achieved with the existing mechanisms. You can modify request by customizing http.Client as follows.

```go
package main

import (
	"net/http"

	"github.com/sashabaranov/go-openai"
)

type CustomRequestTransport struct {
	Origin http.RoundTripper
}

func (t *CustomRequestTransport) RoundTrip(req *http.Request) (*http.Response, error) {
	// Modify request here
	return t.Origin.RoundTrip(req)
}

func main() {
	cfg := openai.DefaultConfig("openai_api_key")
	cfg.HTTPClient = &http.Client{
		Transport: &CustomRequestTransport{
			Origin: http.DefaultTransport,
		},
	}

	client := openai.NewClientWithConfig(cfg)
}

```

Reply:
@vvatanabe Thank you for the code snippet. Custom RoundTripper can do the job.
I'm working on some integration SDK for an API gateway and many of my clients already use their own HTTPClient. It's more cumbersome for them to inject request modification code into `RoundTrip` compared to a dedicate config hook. And the go documentation for RoundTrip recommends against modifying the request.

https://pkg.go.dev/net/http#RoundTripper

> RoundTrip should not modify the request, except for
> consuming and closing the Request's Body.

I see both way can do the job. Would a config hook be more explicit and easier to use?

Reply:
> RoundTrip should not modify the request, except for consuming and closing the Request's Body.

@arrix When interpreting the description of RoundTripper strictly, it can be said that modifying any part of the request, including the headers, is discouraged. However, in practical operations, it's quite common to add or modify headers for specific purposes, such as authentication or tracing.

Generally, when implementing a custom RoundTripper and there's a need to alter headers, it's recommended to clone the original request before making the modifications. This approach prevents changes to the original request object and ensures operations are carried out without impacting other middleware or code.

In summary, while it's fundamentally discouraged to directly modify headers using RoundTripper, this constraint can be bypassed by cloning the request and then making the necessary changes.